compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -952,6 +952,41 @@ where
                 }
             }
 
+            ty::Alias(ty::Opaque, _) => {
+                // This merits some explanation.
+                //
+                // We choose to treat all opaque types as non-local, even
+                // those that appear within the same crate. This seems
+                // somewhat surprising at first, but makes sense when
+                // you consider that opaque types are supposed to hide
+                // the underlying type *within the same crate*. When an
+                // opaque type is used from outside the module
+                // where it is declared, it should be impossible to observe
+                // anything about it other than the traits that it implements.
+                //
+                // The alternative would be to look at the underlying type
+                // to determine whether or not the opaque type itself should
+                // be considered local. However, this could make it a breaking change
+                // to switch the underlying ('defining') type from a local type
+                // to a remote type. This would violate the rule that opaque
+                // types should be completely opaque apart from the traits
+                // that they implement, so we don't use this behavior.
```

Comment:
that part is also somoewhat incomplete: revealing the underlying type is likely to cause cycle errors as we rely on coherence/the specialization graph during typeck